torch/distributed/tensor/experimental/_attention.py

```diff
@@ -1068,27 +1068,37 @@ def _context_parallel_buffers(
     for buffer, seq_dim in zip(buffers, buffer_seq_dims):
         if isinstance(buffer, torch.Tensor):
             # TODO: the load balance doesn't perform error handling.
+
+            # NOTE: assuming batch dim is 0
+
             if load_balance_indices is not None:
-                # NOTE: assuming batch dim is 0
+                # TODO: we should expclitly ask users to unsqueeze the batch dim.
+                # But this is a BC breaking ask.
+                # However, what we have done today is also not very safe.
                 idx_batch_size = load_balance_indices.size(0)
-                data_batch_size = buffer.size(0)
+                data_batch_size = buffer.size(0) if seq_dim > 0 else 1
+
                 if idx_batch_size != 1 and idx_batch_size != data_batch_size:
                     raise ValueError(
                         "Cannot rearrange buffer: "
                         f"load_balance_indices has shape {load_balance_indices.shape}, "
                         f"but buffer has shape {buffer.shape}."
                     )
 
-                for i in range(data_batch_size):
-                    index = (
-                        load_balance_indices[0]  # identical load-balance in batch
-                        if idx_batch_size == 1
-                        else load_balance_indices[i]
-                    )
-                    buffer_batch_i = torch.index_select(
-                        buffer[i], dim=seq_dim - 1, index=index
+                if seq_dim == 0:
+                    buffer = torch.index_select(
+                        buffer, dim=0, index=load_balance_indices[0]
```

Comment:
alternatively, we can just `unsqueeze` inside this function

Reply:
I tried that, but we need to squeeze back before returning. This if/else is more straightforward to me.